### PR TITLE
group mob session

### PR DESCRIPTION
**Pull request by bkd705**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new type `ApplyAlteration` for enhanced type safety in handling alterations.
	- Enhanced `EditToast` and `PlayToast` components with new state management for alterations.

- **Bug Fixes**
	- Improved error management and robustness in asynchronous handling of updates.

- **Improvements**
	- Simplified component logic and enhanced the editing experience by refining state management and function signatures.

- **Chores**
	- Removed outdated functions to streamline component logic and improve performance.

### CODE REVIEW

File: apps/mocksi-lite/background.ts [hunk 1]
```diff
@@ -310,7 +310,7 @@ async function createDemo(body: Record<string, unknown>) {
 
 function updateDemo(data: Record<string, unknown>) {
 	const { id, recording } = data;
```
Comment (fitzk) [inline apps/mocksi-lite/background.ts:312]: nit: maybe we should add a comment here explaining why we are returning the promise?

File: apps/mocksi-lite/content/EditMode/actions.ts [hunk 2]
```diff
@@ -18,14 +18,15 @@ export function applyChanges(
...
+	applyAlteration: ApplyAlteration,
 ) {
 	if (nodeWithTextArea) {
 		cancelEditWithoutChanges(nodeWithTextArea);
 		// TODO: check if we should keep the singleton behavior we had before
 		const domManipulator = new DOMManipulator(
 			fragmentTextNode,
 			getHighlighter(),
-			saveModification,
+			applyAlteration, // TODO: replace this by communicating back into react
```
Comment (fitzk) [inline apps/mocksi-lite/content/EditMode/actions.ts:29]: nit: make a ticket and remove comment?

File: apps/mocksi-lite/commands/Command.ts
Comment (elg0nz) [inline apps/mocksi-lite/commands/Command.ts:1]: YESS! Nothing makes me happier than deleting dead code

File: apps/mocksi-lite/content/AppStateContext.tsx [hunk 3]
```diff
@@ -143,11 +143,6 @@ export const AppStateProvider: React.FC<{
 						payload: AppState.PLAY,
 					});
 					sendMessage("updateToPauseIcon");
```
Comment (elg0nz) [inline apps/mocksi-lite/content/AppStateContext.tsx:145]: Beautiful, so much simpler! ✨

File: apps/mocksi-lite/content/EditMode/actions.ts [hunk 4]
```diff
@@ -81,7 +82,7 @@ export function applyImageChanges(
 	const domManipulator = new DOMManipulator(
 		fragmentTextNode,
 		getHighlighter(),
-		saveModification,
+		() => {},
```
Comment (bkd705) [inline apps/mocksi-lite/content/EditMode/actions.ts:85]: this was previously the saveModification, eventually we'd want to remove this being a required argument I think but this is an okay stop-gap

File: apps/mocksi-lite/content/EditMode/highlighter.ts [hunk 5]
```diff
@@ -126,15 +126,11 @@ const highlight = ({
...
-		highlightedElement.parentElement?.replaceChild(
-			decorate(highlightedElement.textContent || "", `${width || ""}`, false, {
-				onSubmit: undefined,
-				onCancel: () => {
-					(event.target as HTMLElement).style.display = "block";
-				},
-			}),
-			highlightedElement,
-		);
+		// TODO: Come back to handle double clicking on a highlight
```
Comment (fitzk) [inline apps/mocksi-lite/content/EditMode/highlighter.ts:129]: we should dev make a ticket for this I think

File: apps/mocksi-lite/content/ListPopup/DemoItem.tsx [hunk 6]
```diff
@@ -26,13 +25,13 @@ const DemoItem = ({
 	const domain = new URL(url).hostname;
 
 	const handleEdit = async () => {
+		console.log("handleEdit", alterations);
```
Comment (fitzk) [inline apps/mocksi-lite/content/ListPopup/DemoItem.tsx:28]: remove log

Comment (bkd705) [inline apps/mocksi-lite/content/ListPopup/DemoItem.tsx:28]: remove this

File: apps/mocksi-lite/content/Toast/EditToast.tsx [hunk 7]
```diff
@@ -1,28 +1,209 @@
...
+			.catch((err) => {
+				console.error("error fetching alterations", err);
+			});
+	}, []);
+
+	const setupEditor = async () => {
+		sendMessage("attachDebugger");
+
+		observeUrlChange(() => {
+			console.log("URL changed, turning off highlights");
```
Comment (fitzk) [inline apps/mocksi-lite/content/Toast/EditToast.tsx:98]: remove or console.debug

File: apps/mocksi-lite/content/Toast/EditToast.tsx [hunk 8]
```diff
@@ -1,28 +1,209 @@
...
+
+		document.body.removeEventListener("dblclick", onDoubleClickText);
+	};
+
+	const onDoubleClickText = useCallback((event: MouseEvent) => {
+		// @ts-ignore MouseEvent typing seems incomplete
+		const nodeName = event?.toElement?.nodeName;
+		console.log("we double clicked on", event.target);
+
+		// if (nodeName === "IMG") {
```
Comment (fitzk) [inline apps/mocksi-lite/content/Toast/EditToast.tsx:160]: make ticket to add this back when we know it's working

File: apps/mocksi-lite/content/EditMode/editMode.ts [hunk 9]
```diff
@@ -1,116 +1,8 @@
...
-	sendMessage,
-	undoModifications,
-} from "../../utils";
-import { applyImageChanges, cancelEditWithoutChanges } from "./actions";
+import { MOCKSI_READONLY_STATE } from "../../consts";
+import type { ApplyAlteration } from "../Toast/EditToast";
+import { applyImageChanges } from "./actions";
 import { decorate } from "./decorator";
 
-const observeUrlChange = (onChange: () => void) => {
```
Comment (elg0nz) [inline apps/mocksi-lite/content/EditMode/editMode.ts:12]: Nice!!

File: apps/mocksi-lite/content/EditMode/highlighter.ts (hunk 5 above)
Comment (elg0nz) [inline apps/mocksi-lite/content/EditMode/highlighter.ts:129]: nit: might be better to file a ticket instead

File: apps/mocksi-lite/content/ListPopup/DemoItem.tsx (hunk 6 above)
Comment (elg0nz) [inline apps/mocksi-lite/content/ListPopup/DemoItem.tsx:28]: nit: use `console.debug` instead

File: apps/mocksi-lite/content/content.tsx [hunk 10]
```diff
@@ -16,43 +16,26 @@ import {
...
-		const appState: AppState | null = results[MOCKSI_RECORDING_STATE];
-		if (appState === AppState.PLAY) {
-			handlePlayState();
-		}
-
-		if (appState === AppState.EDITING) {
-			handleEditState();
-		}
-	});
+	// TODO: explore if we can auto open extension for hard navigation sites
```
Comment (fitzk) [inline apps/mocksi-lite/content/content.tsx:28]: nit: I think we can remove this 

Maybe check for a ticket for the hard reload issue and if it doesn't exist make one

File: apps/mocksi-lite/utils.ts [hunk 11]
```diff
@@ -250,21 +229,21 @@ export const loadAlterations = async (
...
 // This is from chrome.storage.local
-export const loadPreviousModifications = () => {
-	for (const [
-		querySelector,
-		{ oldValue, newValue, type },
-	] of modificationsIterable()) {
-		const sanitizedOldValue = sanitizeHtml(oldValue);
-		const elemToModify = getHTMLElementFromSelector(querySelector);
+// this should be called "revertModifications"
+export const loadPreviousModifications = (alterations: Alteration[]) => {
```
Comment (fitzk) [inline apps/mocksi-lite/utils.ts:233]: nit: I would be so happy if we renamed this fn haha

File: packages/dodom/receivers/DOMManipulator.ts [hunk 12]
```diff
@@ -61,6 +61,7 @@ export class DOMManipulator {
 		this.iterateAndReplace(body, pattern, newText, true);
 	}
 
+	// Cash -> Gold
```
Comment (fitzk) [inline packages/dodom/receivers/DOMManipulator.ts:64]: nit: lmao I say we keep this comment 😜

File: apps/mocksi-lite/content/Toast/EditToast.tsx (hunk 8 above)
Comment (elg0nz) [inline apps/mocksi-lite/content/Toast/EditToast.tsx:160]: FYI @fitzk I just assigned you https://linear.app/mocksi/issue/MOC-73/highlight-image-changes which will require you restoring this code

File: packages/dodom/receivers/DOMManipulator.ts (hunk 12 above)
Comment (elg0nz) [inline packages/dodom/receivers/DOMManipulator.ts:64]: this seems like an extra comment

File: apps/mocksi-lite/content/EditMode/actions.ts (hunk 4 above)
Comment (fitzk) [reply apps/mocksi-lite/content/EditMode/actions.ts:85]: for sure, maybe a comment explaining we should remove when we update how DomMan... is instantiated

**Review by elg0nz — COMMENTED** [review]
Couple small nits, but overall this is 🔥 🔥 
Can't wait to get a build with this stuff in

File: apps/mocksi-lite/content/EditMode/actions.ts (hunk 2 above)
Comment (bkd705) [reply apps/mocksi-lite/content/EditMode/actions.ts:29]: this was actually done as part of our work! was an old commend when it was still saveModification! removed the comment tho

File: apps/mocksi-lite/content/EditMode/actions.ts [hunk 13]
```diff
@@ -18,14 +18,15 @@ export function applyChanges(
...
+	applyAlteration: ApplyAlteration,
 ) {
 	if (nodeWithTextArea) {
 		cancelEditWithoutChanges(nodeWithTextArea);
 		// TODO: check if we should keep the singleton behavior we had before
 		const domManipulator = new DOMManipulator(
 			fragmentTextNode,
 			getHighlighter(),
-			saveModification,
+			applyAlteration,
```
Comment (elg0nz) [inline apps/mocksi-lite/content/EditMode/actions.ts:29]: yay

File: apps/mocksi-lite/content/Toast/EditToast.tsx [hunk 14]
```diff
@@ -1,28 +1,205 @@
...
 };
 
+export type ApplyAlteration = (
+	element: HTMLElement,
+	newText: string,
+	cleanPattern: string,
+	type: "text" | "image",
+) => void;
+
+const observeUrlChange = (onChange: () => void) => {
```
Comment (elg0nz) [inline apps/mocksi-lite/content/Toast/EditToast.tsx:41]: BOOM! 💥

File: apps/mocksi-lite/utils.ts [hunk 15]
```diff
@@ -72,72 +67,58 @@ export const logout = () => {
...
 		updated_timestamp,
 		alterations,
 	});
-	sendMessage("updateDemo", {
-		id: recordingId,
-		recording: { updated_timestamp, alterations },
+
+	// Return a promise here so we can "await" the response
+	// This allows us to ensure the demo has been updated before taking the next action (ie. closing the editor)
+	return new Promise((resolve) => {
```
Comment (elg0nz) [inline apps/mocksi-lite/utils.ts:83]: Niice

**Review by elg0nz — APPROVED** [review]
🚢 🚀 :shipit: :all-the-ship-related-emojis-available-all-at-once: 💥

File: apps/mocksi-lite/content/EditMode/highlighter.ts (hunk 5 above)
Comment (fitzk) [reply apps/mocksi-lite/content/EditMode/highlighter.ts:129]: https://linear.app/mocksi/issue/MOC-174/fix-double-click-on-highlight

File: apps/mocksi-lite/content/content.tsx (hunk 10 above)
Comment (fitzk) [reply apps/mocksi-lite/content/content.tsx:28]: https://linear.app/mocksi/issue/MOC-175/handle-websites-with-hard-reload-on-navigation

**Pull request merged**